### ISSUE TITLE
Change Request: Bulk Suppressions: Resolve `eslint-suppressions.json` relative to linted files

### ISSUE THREAD

**Issue by Grapedge** [eslint/eslint#19954]
### ESLint version

v9.30.1

#### Context
We are migrating a **Rush-based monorepo** (150+ packages, 1M+ TS LoC, multi-team ownership) from ESLint v8 to v9. Our project structure is:
```
├── apps/
│   ├── app1/
│   │   ├── eslint.config.mjs   # Per-package config
│   ├── app2/
│       ├── eslint.config.mjs
├── packages/
│   ├── team-a/
│   │   ├── lib1/
│   │   │   ├── eslint.config.mjs
│   │   ├── lib2/
│   │       ├── eslint.config.mjs
│   ├── team-b/
│       ├── lib3/
│           ├── eslint.config.mjs
└── rush.json                   # Monorepo root
```

We use **multiple ESLint configs** via `--flag v10_config_lookup_from_file` (per [discussion #16960](https://github.com/eslint/eslint/discussions/16960)), which works well.

#### Problem: Inconsistent Suppression File Resolution
During migration, we need to:
1. Replace obsolete `eslint-disable` comments with new rules
2. Apply **Bulk Suppressions** for newly enabled rules

But current `eslint-suppressions.json` lookup **fails in multi-root scenarios** due to different working directories:

| Execution Context          | Command Example                                                                 | Current Lookup Path               | Desired Lookup Path              |
|----------------------------|---------------------------------------------------------------------------------|-----------------------------------|----------------------------------|
| **Developer Local**        | `cd packages/lib1 && rushx lint`                                                | `packages/lib1/` (correct)        | (Same)                           |
| **CI/Lint-Staged**         | `eslint --flag v10_config_lookup_from_file packages/lib1/src/foo.ts packages/lib2/src/bar.ts apps/app1/dev.ts` | Monorepo root (❌ fails) | Respective package directories |

This causes:
- ❌ CI cannot find package-specific suppressions
- ❌ Manual workarounds required for cross-package linting

### What do you think is the correct solution?

**Align `eslint-suppressions.json` resolution with ESLint config resolution:**  
Make it respect `v10_config_lookup_from_file`'s behavior by resolving suppression files **relative to linted files**, mirroring how ESLint finds configs:
```
├── apps/
│   ├── app1/
│   │   ├── eslint.config.mjs
│   │   ├── eslint-suppressions.json   ✅ Resolved when linting app1 files
├── packages/
│   ├── team-a/
│   │   ├── lib1/
│   │   │   ├── eslint.config.mjs
│   │   │   ├── eslint-suppressions.json   ✅ Resolved when linting lib1 files
```

### Participation

- [ ] I am willing to submit a pull request for this change.

### Additional comments

- Already using experimental flags (`--flag v10_config_lookup_from_file`) and `--pass-on-unpruned-suppressions`

**Comment by nzakas** [comment]
This is interesting. I can't quite tell from your description what is going on. Can you explain how you're running ESLint? Do you run it once across all packages, or once per package?

**Comment by Grapedge** [comment]
> Do you run it once across all packages, or once per package?

We have two execution approaches:

1. **Per-package execution**: Enter each package's directory and run `eslint .`  
   ```bash
   cd packages/bar
   pnpm lint  # eslint ./ --cache
   ```

2. **Integration with tools like lint-staged**: lint-staged passes files from the git staging area directly to ESLint. For example, when modifying both `packages/bar/index.js` and `packages/foo/index.js`, it executes:
  ```bash
  eslint --fix --flag v10_config_lookup_from_file --prune-suppressions --no-warn-ignored /Users/xxx/Documents/eslint-suppressions-monorepo/packages/bar/index.js /Users/xxx/Documents/eslint-suppressions-monorepo/packages/foo/index.js
  ```

For clearer illustration, I've created a demo repository: [eslint-suppressions-monorepo](https://github.com/Grapedge/eslint-suppressions-monorepo)

**Comment by nzakas** [comment]
Thanks, that's helpful.

Given the lint-staged use case, it seems like you'd be better off keeping the suppressions file at the root of your project and always running ESLint from there.

I don't think it's practical to dynamically determine the suppressions file based on the file being linted. That would mean any given run on ESLint might need to know about *n* suppression files, and then we'd have to worry about things like conflicts and how we keep the suppressions files up-to-date.

**Comment by Grapedge** [comment]
I previously attempted placing it in the root directory—this file totaled around 120,000 lines. Since I wanted `eslint-suppressions.json` to stay as up-to-date as possible by having everyone run prune suppressions during git commit, I anticipated a high probability of merge conflicts. Splitting it at the package level reduces this conflict risk.

Another reason I ultimately didn’t place it in the root directory is that it would disrupt developers' workflow habits. Team members typically focus only on the packages they work on, and some may use VSCode’s [multi-root workspaces](https://code.visualstudio.com/docs/editing/workspaces/multi-root-workspaces). In those cases, the root directory doesn’t functionally exist for them.

<img width="898" height="319" alt="Image" src="https://github.com/user-attachments/assets/8613831b-7221-4b6c-ba12-175fc5726f0d" />

Our current solution handles this in `lint-staged.config.js`: We identify the owning package for each staged file, group files by their respective packages, then cd into each package directory to execute ESLint.

> That would mean any given run on ESLint might need to know about n suppression files

Perhaps we could establish a convention to place `eslint-suppressions.json` alongside `eslint.config.js`? I'm uncertain whether this is a practical approach.

**Comment by nzakas** [comment]
> Perhaps we could establish a convention to place `eslint-suppressions.json` alongside `eslint.config.js`? I'm uncertain whether this is a practical approach.

Because each file can use a different `eslint.config.js`, that doesn't really work. You'd end up with the same multiple `eslint-suppressions.json` problem.

The only suggestion I have is to run ESLint separately for each package, so each package can have its own `eslint-suppressions.json` file. Then modify the top-level linting script to run and aggregate results from each package.

**Comment by privatenumber** [comment]
We're also hitting this limitation at Square.

In our large monorepo, multiple teams work in their own packages, each with its own suppression file to track tech debt. A single root-level suppression file would make it much harder to hold teams accountable for their own areas.

I think this ties into the broader RFC in https://github.com/eslint/eslint/issues/19978 about suppressed error reminders. If suppression files could be resolved per file (like configs) and kept within their respective directories, it'd be much easier to maintain clear ownership and manage tech debt in a big repo.
